### PR TITLE
Feature mpi pack

### PR DESCRIPTION
**Pull request by lukasdreyer**
# MPI Pack/Unpack

Proposed changes:
For t8code, we are interested to use MPI_Pack, MPI_Unpack and MPI_Pack_size. This PR introduces a wrapper for these functionalities when libsc is compiled without MPI.

### CODE REVIEW

**Comment by cburstedde** [comment]
Thanks a lot for giving back to sc! Appreciate the update.

Might I ask you to (a) provide a doc/author_dreyer.txt file like the others, (b) provide the usual doxygen comments in the .h and if you have time (c) give the function some minimal test coverage?

**Comment by lukasdreyer** [comment]
> Thanks a lot for giving back to sc! Appreciate the update.
> 
> Might I ask you to (a) provide a doc/author_dreyer.txt file like the others, (b) provide the usual doxygen comments in the .h and if you have time (c) give the function some minimal test coverage?

It's a pleasure. Writing a test even helped uncover an error that was still included with multiple elements that were packed.

File: src/sc_mpi.c [hunk 1]
```diff
@@ -914,3 +916,42 @@ sc_mpi_comm_get_and_attach (sc_MPI_Comm mpicomm)
...
+{
+  int                 size;
+  sc_MPI_Pack_size (incount, datatype, comm, &size);
+  /** Check that we have enough space to pack the datatypes */
+  if (*position + size > outsize)
+    return sc_MPI_ERR_NO_SPACE;
+  /* Copy the contiguous memory */
+  memcpy ((char *) outbuf + *position, inbuf, size);
+  *position += size;
+  return 0;
```
Comment (cburstedde) [inline src/sc_mpi.c:932]: We like to return sc_MPI_SUCCESS throughout.

File: src/sc_mpi.c [hunk 2]
```diff
@@ -914,3 +916,42 @@ sc_mpi_comm_get_and_attach (sc_MPI_Comm mpicomm)
 
   return intrasize;
 }
+
+int
+sc_MPI_Pack (const void *inbuf, int incount, sc_MPI_Datatype datatype,
+             void *outbuf, int outsize, int *position, sc_MPI_Comm comm)
+{
+  int                 size;
+  sc_MPI_Pack_size (incount, datatype, comm, &size);
```
Comment (cburstedde) [inline src/sc_mpi.c:925]: We usually test all MPI functions' returns with SC_CHECK_MPI (mpiret).

File: src/sc_mpi.c [hunk 3]
```diff
@@ -914,3 +916,42 @@ sc_mpi_comm_get_and_attach (sc_MPI_Comm mpicomm)
...
+  *position += size;
+  return 0;
+}
+
+int
+sc_MPI_Pack_size (int incount, sc_MPI_Datatype datatype, sc_MPI_Comm comm,
+                  int *size)
+{
+  sc_MPI_Type_size (datatype, size);
+  *size *=incount;
```
Comment (cburstedde) [inline src/sc_mpi.c:956]: If you could run scindent on the code (the prototype spacing does not need to conform (indent is not 100% consistent), but the spacing of the C code itself benefits from it)?

**Comment by cburstedde** [comment]
Thanks again! I'm wondering whether you might look at the CI to see what's up.

File: src/sc_mpi.c (hunk 3 above)
Comment (lukasdreyer) [reply src/sc_mpi.c:956]: When I run scindent on sc_mpi.c, I get lots of changes of the following form:
```
-               sc_MPI_Status * status)
+               sc_MPI_Status *status)
```
Should I include them in this PR as well?

**Comment by cburstedde** [comment]
> When I run scindent on sc_mpi.c, I get lots of changes of the following form:
> ```
> -               sc_MPI_Status * status)
> +               sc_MPI_Status *status)
> ```
> Should I include them in this PR as well?

I know, these are annoying.  Please ignore those.  Thanks!

**Comment by lukasdreyer** [comment]
> When I run scindent on sc_mpi.c, I get lots of changes of the following form: ``` - sc_MPI_Status * status) + sc_MPI_Status *status) ``` Should I include them in this PR as well?
> I know, these are annoying. Please ignore those. Thanks!

Good to know. Should be ready for the workflow check again

**Comment by cburstedde** [comment]
> > When I run scindent on sc_mpi.c, I get lots of changes of the following form: ``` - sc_MPI_Status * status) + sc_MPI_Status *status) ``` Should I include them in this PR as well?
> > I know, these are annoying. Please ignore those. Thanks!
> 
> Good to know. Should be ready for the workflow check again

Thanks; I've just pushed some minor convention alignments.

One more thing: The three new functions are currently in a section of the .c
file where they are compiled even with MPI enabled.  Would you please move
them up to the MPI replacement functions that are only compiled without MPI?

File: test/test_mpi_pack.c [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+}
+test_message_t;
+
+void
+test_message_construct (test_message_t *message, int8_t type,
+                        double startvalue)
+{
+  message->type = type;
+  message->values = SC_ALLOC (double, num_values[type]);
+  for (int i = 0; i < num_values[type]; i++) {
```
Comment (cburstedde) [inline test/test_mpi_pack.c:49]: We stick to the old C standard where declarations must come right at the beginning of a block before any code.

File: test/test_mpi_pack.c [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+/* Pack several instances of test messages into contiguous memory */
+int
+test_message_MPI_Pack (const test_message_t *messages, int incount,
+                       void *outbuf, int outsize, int *position,
+                       sc_MPI_Comm comm)
+{
+  for (int imessage = 0; imessage < incount; imessage++) {
+    sc_MPI_Pack (&(messages[imessage].type), 1, sc_MPI_INT8_T, outbuf,
```
Comment (cburstedde) [inline test/test_mpi_pack.c:81]: Would you mind doing a SC_CHECK_MPI (mpiret) for all MPI calls?

File: test/test_mpi_pack.c [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+/* Determine how much space in bytes is needed to pack several test messages */
+int
+test_message_MPI_Pack_size (int incount, const test_message_t *messages,
+                            sc_MPI_Comm comm, int *size)
+{
+  *size = 0;
+  for (int imessage = 0; imessage < incount; imessage++) {
+    int                 pack_size;
```
Comment (cburstedde) [inline test/test_mpi_pack.c:114]: This is cool, beginning of a block.

File: test/test_mpi_pack.c [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  int                 mpiret;
+  sc_MPI_Comm         mpicomm;
+
+  /* standard initialization */
+  mpiret = sc_MPI_Init (&argc, &argv);
+  SC_CHECK_MPI (mpiret);
+  mpicomm = sc_MPI_COMM_WORLD;
+  sc_init (mpicomm, 1, 1, NULL, SC_LP_DEFAULT);
+
+  int                 num_test_messages = 5;
```
Comment (cburstedde) [inline test/test_mpi_pack.c:138]: These and further declarations please move to the top before any code is executed.

**Comment by cburstedde** [comment]
Thanks again; just left a few messages about the test program. Then ready to merge!

**Comment by lukasdreyer** [comment]
> Thanks again; just left a few messages about the test program. Then ready to merge!
Done

**Comment by cburstedde** [comment]
Thanks your contribution and for following up on the PR!

**Comment by cburstedde** [comment]
(Sorry misclick. About to merge.)

**Comment by cburstedde** [comment]
Just checking: how important is MPI_INT8_T for y'all? It's from a relatively new standard version. We'd rather only use MPI up to 1.3. Would you be fine removing that type and also MPI_UNSIGNED_LONG_LONG? Please @hannesbrandt chime in for more details/comments.

**Pull request merged**